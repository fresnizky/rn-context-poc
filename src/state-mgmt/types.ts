import { IArtistState } from './artist/state';
import { IAlbumState } from './album/state';

export interface IAction<T = string, P = any> {
  type: T;
  payload: P;
}

export type IGlobalState = {
  artist: IArtistState;
  album: IAlbumState;
};

export interface Deps {
  apiService: { request: Function };
  stateSnapshot: { get: () => IGlobalState; set: (state: IGlobalState) => void };
}
