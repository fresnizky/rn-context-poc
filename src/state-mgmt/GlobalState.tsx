import React, { createContext, memo, PropsWithChildren, useReducer, Dispatch, useEffect } from 'react';

import { initialState as artistInitialState } from './artist/state';
import { initialState as albumInitialState } from './album/state';
import { reducer as artistReducer } from './artist/reducer';
import { reducer as albumReducer } from './album/reducer';
import { IAction, IGlobalState, Deps } from './types';

export const initialState: IGlobalState = {
  artist: artistInitialState,
  album: albumInitialState
};

export function combineReducers<GS>(reducerMap: Record<string, (state: any, action: IAction) => any>) {
  return (gState: GS, gAction: IAction) => {
    const nextState = Object.entries(reducerMap).reduce(
      (total, [stateKey, reducer]) => ({ ...total, [stateKey]: reducer(gState[stateKey], gAction) }),
      {} as GS
    );
    const hasChanged = !!Array.from(new Set([...Object.keys(gState), ...Object.keys(nextState)])).find(key => gState[key] !== nextState[key]);
    return hasChanged ? nextState : gState;
  };
}

export const GlobalContext = createContext<{ state: IGlobalState; dispatch: Dispatch<any>; deps: Deps }>({
  state: initialState,
  dispatch: () => null,
  deps: null
});

export const GlobalProvider = memo(({ children, deps, initState = initialState }: PropsWithChildren<{ deps: Deps; initState?: IGlobalState }>) => {
  const [state, dispatch] = useReducer(
    combineReducers<IGlobalState>({ artist: artistReducer, album: albumReducer }),
    initState
  );
  useEffect(() => deps.stateSnapshot.set(state), [state]);
  return <GlobalContext.Provider value={{ state, dispatch, deps }}>{children}</GlobalContext.Provider>;
});
