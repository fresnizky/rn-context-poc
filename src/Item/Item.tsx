import React, { memo, useContext, useEffect } from 'react';
import { ScrollView, Text, Image, View, FlatList } from 'react-native';
import { useAlbumEffect } from '../state-mgmt/album/useAlbumEffect';
import { GlobalContext } from '../state-mgmt/GlobalState';

export interface Props {
  route: any; // RouteProp<{ id: string; }, any>; /** @todo find out how to use this type */
}

const Item = ({ route }: Props) => {
  const { state } = useContext(GlobalContext);
  console.log(state);
  const artist = state.artist.artistMap[route?.params?.id];
  const albums = state.album;
  const { searchAlbumByArtistId } = useAlbumEffect();

  useEffect(() => {
    searchAlbumByArtistId(artist);
  }, [artist.idArtist]);

  return (
    <ScrollView contentInsetAdjustmentBehavior="automatic">
      <Image style={{ width: '100%', height: 100 }} source={{ uri: artist.strArtistBanner }} />
      <View style={{ padding: 15 }}>
        <Text>{artist.strArtist}</Text>
        <Text ellipsizeMode="head" numberOfLines={30}>
          {artist.strBiographyEN}
        </Text>
      </View>
      {albums && (
        <View>
          <Text>Albums:</Text>
          <ScrollView horizontal>
            {Object.keys(albums).map(albumId =>
              albums[albumId].idArtist === artist.idArtist && albums[albumId].strAlbumThumb ? (
                <Image key={albumId} style={{ width: 50, height: 50 }} source={{ uri: albums[albumId].strAlbumThumb }} />
              ) : null
            )}
          </ScrollView>
        </View>
      )}
    </ScrollView>
  );
};

export default memo(Item);
