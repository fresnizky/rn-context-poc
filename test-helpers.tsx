import { IArtist } from './src/state-mgmt/artist/state';
import { Deps } from './src/state-mgmt/types';
import { IAlbum } from 'src/state-mgmt/album/state';

export const getMockDeps = (): Deps =>
  ({
    apiService: { request: jest.fn().mockResolvedValue({ json: () => ({}) }) },
    stateSnapshot: { set: jest.fn(), get: jest.fn().mockReturnValue({}) }
  } as any);

export const getArtist_1 = (): IArtist => ({
  idArtist: '1',
  intBornYear: 1970,
  strArtist: 'Freddie Mercuty',
  strArtistThumb: 'thumb.png',
  strArtistBanner: 'banner.png',
  strArtistFanart: 'fanart.png',
  strBiographyEN: 'this is a bio',
  strCountry: 'USA',
  strStyle: 'rock'
});

export const getAlbum_1 = (): IAlbum => ({
  idAlbum: '1',
  idArtist: '1',
  intYearReleased: '2000',
  strAlbum: 'Mr. Bad Guy',
  strAlbumCDart: 'cdart.png',
  strAlbumThumb: 'thumb.png',
  strAlbumThumbBack: 'thumbBack.png',
  strDescriptionEN: 'Album Description',
  strGenre: 'Pop'
});
