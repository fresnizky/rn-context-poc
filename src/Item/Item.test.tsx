import React from 'react';
import { render } from 'react-native-testing-library';

import Item from './Item';
import { getArtist_1, getAlbum_1, getMockDeps } from '../../test-helpers';
import { GlobalProvider, initialState } from '../state-mgmt/GlobalState';

describe('Item', () => {
  it('should render', () => {
    const { toJSON } = render(
      <GlobalProvider
        deps={getMockDeps()}
        initState={{
          ...initialState,
          artist: { ...initialState.artist, artistMap: { [getArtist_1().idArtist]: getArtist_1() } },
          album: { ...initialState.album, albumMap: { [getAlbum_1().idAlbum]: getAlbum_1() } }
        }}
      >
        <Item route={{ params: { id: getArtist_1().idArtist } }} />
      </GlobalProvider>
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
