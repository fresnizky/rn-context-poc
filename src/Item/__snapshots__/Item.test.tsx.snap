// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item should render 1`] = `
<RCTScrollView
  contentInsetAdjustmentBehavior="automatic"
>
  <View>
    <Image
      source={
        Object {
          "uri": "banner.png",
        }
      }
      style={
        Object {
          "height": 100,
          "width": "100%",
        }
      }
    />
    <View
      style={
        Object {
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 24,
            "fontWeight": "bold",
            "lineHeight": 32,
            "marginBottom": 5,
          }
        }
      >
        Freddie Mercuty
      </Text>
      <Text
        ellipsizeMode="head"
        numberOfLines={30}
        style={
          Object {
            "fontSize": 16,
            "lineHeight": 24,
            "textAlign": "justify",
          }
        }
      >
        this is a bio
      </Text>
    </View>
    <View
      style={
        Object {
          "paddingBottom": 10,
          "paddingLeft": 10,
          "paddingRight": 10,
          "paddingTop": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "fontSize": 24,
            "fontWeight": "bold",
            "lineHeight": 32,
            "marginBottom": 5,
          }
        }
      >
        Albums:
      </Text>
      <RCTScrollView
        data={Array []}
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={true}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        style={
          Object {
            "marginRight": -10,
          }
        }
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View />
      </RCTScrollView>
    </View>
  </View>
</RCTScrollView>
`;
